src/sentry/db/models/manager/base.py

```diff
@@ -438,3 +451,54 @@ def get_queryset(self) -> BaseQuerySet:
         if hasattr(self, "_hints"):
             return self._queryset_class(self.model, using=self._db, hints=self._hints)
         return self._queryset_class(self.model, using=self._db)
+
+    def create_mode_limited_copy(
+        self, limit: ModeLimited, read_modes: Iterable[ServerComponentMode]
+    ) -> BaseManager[M]:
+        """Create a copy of this manager that enforces server mode limitations."""
+
+        # Dynamically create a subclass of this manager's class, adding overrides.
+        cls = type(self)
+        overrides = {
+            "get_queryset": limit.create_override(cls.get_queryset, extra_modes=read_modes),
+            "bulk_create": limit.create_override(cls.bulk_create),
+            "bulk_update": limit.create_override(cls.bulk_update),
+            "create": limit.create_override(cls.create),
+            "create_or_update": limit.create_override(cls.create_or_update),
+            "get_or_create": limit.create_override(cls.get_or_create),
+            "post_delete": limit.create_override(cls.post_delete),
+            "select_for_update": limit.create_override(cls.select_for_update),
+            "update": limit.create_override(cls.update),
+            "update_or_create": limit.create_override(cls.update_or_create),
+        }
+        manager_subclass = type(cls.__name__, (cls,), overrides)
+        manager_instance = manager_subclass()
+
+        # Ordinarily a pointer to the model class is set after the class is defined,
+        # meaning we can't inherit it. Manually copy it over now.
+        manager_instance.model = self.model
+
+        # Copy over some more stuff that would be set in __init__
+        # (warning: this is brittle)
+        manager_instance.cache_fields = self.cache_fields
+        manager_instance.cache_ttl = self.cache_ttl
+        manager_instance._cache_version = self._cache_version
+        manager_instance.__local_cache = threading.local()
```

Comment:
Should we add comments to the BaseManager about needing to update this code if any additional manager properties are introduced?

Reply:
Yeah, that's a good idea. It's unfortunately going to be kludgey no matter what -- it breaks some OOP rules in that it can't anticipate additional stuff that might be added in the `__init__` of some other `BaseManager` subclass.